convert/convert_granite.go

```diff
@@ -0,0 +1,231 @@
+package convert
+
+import (
+	"cmp"
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/pdevine/tensor"
+	"github.com/pdevine/tensor/native"
+
+	"github.com/ollama/ollama/fs/ggml"
+)
+
+type graniteModel struct {
+	ModelParameters
+	NLayers               uint32  `json:"n_layers"`
+	NumHiddenLayers       uint32  `json:"num_hidden_layers"`
+	NLayer                uint32  `json:"n_layer"`
+	MaxPositionEmbeddings uint32  `json:"max_position_embeddings"`
+	NCtx                  uint32  `json:"n_ctx"`
+	HiddenSize            uint32  `json:"hidden_size"`
+	NEmbd                 uint32  `json:"n_embd"`
+	IntermediateSize      uint32  `json:"intermediate_size"`
+	NInner                uint32  `json:"n_inner"`
+	NumAttentionHeads     uint32  `json:"num_attention_heads"`
+	NHead                 uint32  `json:"n_head"`
+	NumKeyValueHeads      uint32  `json:"num_key_value_heads"`
+	RopeTheta             float32 `json:"rope_theta"`
+	RopeScaling           struct {
+		Type                            string  `json:"type"`
+		RopeType                        string  `json:"rope_type"`
+		Factor                          float32 `json:"factor"`
+		LowFrequencyFactor              float32 `json:"low_freq_factor"`
+		HighFrequencyFactor             float32 `json:"high_freq_factor"`
+		OriginalMaxPositionalEmbeddings uint32  `json:"original_max_positional_embeddings"`
+
+		factors ropeFactor
+	} `json:"rope_scaling"`
+	RMSNormEPS          float32 `json:"rms_norm_eps"`
+	LayerNormEPS        float32 `json:"layer_norm_eps"`
+	LayerNormEpsilon    float32 `json:"layer_norm_epsilon"`
+	NormEpsilon         float32 `json:"norm_epsilon"`
+	HeadDim             uint32  `json:"head_dim"`
+	AttentionMultiplier float32 `json:"attention_multiplier"`
+	EmbeddingMultiplier float32 `json:"embedding_multiplier"`
+	ResidualMultiplier  float32 `json:"residual_multiplier"`
+	LogitsScaling       float32 `json:"logits_scaling"`
+}
+
+var _ ModelConverter = (*graniteModel)(nil)
+
+func (p *graniteModel) KV(t *Tokenizer) ggml.KV {
+	kv := p.ModelParameters.KV(t)
+	kv["general.architecture"] = "granite"
+	kv["granite.vocab_size"] = p.VocabSize
+
+	kv["granite.block_count"] = cmp.Or(p.NLayers, p.NumHiddenLayers, p.NLayer)
```

Comment:
Matt pointed out to me that we probably don't need these extra aliases for the layer count

Reply:
Actually, I have a complete, per-field break-down and we should discuss an approach (intelligent commenting) which allows future devs. to understand the delta.